### ISSUE TITLE
Move math from pyro.ops into funsor

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/funsor#362]
During past projects that have used Pyro, I've tried to publish any reusable math components in [pyro.ops](http://docs.pyro.ai/en/stable/ops.html). In the future I'd like to publish these somewhere in Funsor so they are available in both Pyro and NumPyro. For example I'd like to move Pyro's [HaarTransform](http://docs.pyro.ai/en/stable/distributions.html#pyro.distributions.transforms.HaarTransform) into Funsor, unblocking `numpyro.contrib.epidemiology`.

This issue proposes reorganizing `funsor.ops` to contain all backend-generic low-level ops as follows:
```
funsor/ops/__init__.py  # imports Op and basic ops
funsor/ops/op.py  # defines Op
funsor/ops/builtin.py  # defines basic ops like add, mul, etc
funsor/ops/array.py  # defines things like GetitemOp, cat, stack, reshape
funsor/ops/transforms.py  # defines Haar wavelets, DCT transform
funsor/ops/phylo.py  # defines phylogenetic likelihoods
...
```
This module would be the main interface for ops, and the backend-specific modules like `funsor.torch.ops` would simply register implementations of those interfaces. We can organize `funsor.mybackend.ops` as either a monolithic module or as multiple submodules; it doesn't matter because it isn't an interface.

cc @eb8680 @fehiepsi

**Comment by eb8680** [comment]
This seems like a good idea to me. What other ops should we move from Pyro and NumPyro versus eventual deprecation in favor of existing Funsor code?  What about `pyro.ops.gaussian`/`pyro.ops.gamma_gaussian`?

**Comment by fritzo** [comment]
>What other ops should we move?

Probably most of pyro.ops.tensor_utils, .arrowhead, .hessian, .linalg, .newton. Maybe .integrator, .dual_averaging. But I think we can move as-needed.

> What about pyro.ops.gaussian

I think pyro.ops.gaussian is basically already ported to funsor.gaussian, but it would be good to profile speed before deprecating the Pyro versions.

> pyro.ops.gamma_gaussian?

As needed.

**Comment by fritzo** [comment]
Fixed by #363
